docs/ql-style-guide.md

```diff
@@ -0,0 +1,434 @@
+  int linesOfCode, int nestingDepth, int numberOfBranches) {
+  ...
+}
+```
+	
+```ql
+from Method main
+where main.getName() = "Main"
+select main, "This is the program entry point." 
+```
+
+```ql
+from Method main
+where main.getName() = "Main"
+select main, "This is the program entry point." 
+```
+
+```ql
+from Method main
+where
+  main.getName() = "Main" and
+  main.getNumberOfParameters() = 0
+select main, "Main method has no parameters."
+```	
+
+```ql
+  if x.isPublic()
+  then result = "public"
+  else result = "private"
+```	
+
+```ql
+  if
+    x.isPublic()
+  then
+    result = "public"
+  else
+    result = "private"
+```
+
+## Braces
+1. Braces follow "Stroustrup" style: The opening `{` *must* be placed at the end of the preceding line.
+1. The closing `}` *must* be placed on its own line, indented to the outer level, or be on the same line as the opening `{`.
+1. Braces of empty blocks *may* be placed on a single line, with a single space separating the braces.
+1. Short predicates, not exceeding the maximum line width, *may* be placed on a single line, with a space following the opening brace and preceding the closing brace.
+
+
+For example,
+
+```ql
+class ThrowException extends ThrowExpr {
+  Foo() {
+    this.getTarget() instanceof ExceptionClass
+  }
+
+  override string toString() { result = "Throw Exception" }
+}
+```
+
+## Spaces
+1. There *must* be a space or line break:
+   - surrounding each `=` and `|`,
+   - after each comma.
+1. There *should* be a space or line break:
+   - surrounding each *binary operator*, which *must* be balanced,
+   - surrounding `..` in a range.
+   - Exceptions to this are to save space or to improve readability.
+1. *Avoid* other spaces, for example:
+   - after a *quantifier/aggregation* keyword,
+   - after the predicate name in a *call*,
+   - inside brackets used for *calls*, single-line quantifiers, and parenthesised formulas,
+   - surrounding a `.`,
+   - inside the opening or closing `[ ]` In a range expression,
+   - inside casts `a.(X)`.
+1. *Avoid* multiple spaces, except for indentation, and *avoid* additional indentation to align formulas, parameters or arguments.
+1. *Do not* put whitespace on blank lines, or trailing on the end of a line.
+1. *Do not* use tabs.
+
+
+For example,
+
+```ql
+cached
+private predicate foo(Expr e, Expr p) {
+  exists(int n |
+    n in [0 .. 1] |
+    e = p.getChild(n + 1)
+  )
+}
+```
+
+## Naming
+1. Use PascalCase for:
+   - `class` names,
+   - `module` names,
+   - `newtype` names.
+1. Use camelCase for:
+   - predicate names,
+   - variable names.
+1. Newtype predicate names *should* begin with `T`.
+1. *Avoid* underscores in names.
+1. Predicates that have a result *should* be named `get...`
+1. Predicates that can return multiple results *should* be named `getA...` or `getAn...`
+1. Predicates that don't have a result or parameters *should* be named `is...` or `has...`

```

Comment:
This appears to suggest that predicate names should be English.  Is it worth stating our policy on languages (used in names and comments) explicitly?

Reply:
Good idea.